File: objects/calculate-areas.js
/*
Input:
const shapes = [
  {type: 'SQUARE', edge:[4]},
  {type: 'TRIANGLE', edge:[3,4]},
  {type: 'CIRCLE', edge: [1] },
];

Output:
calculateArea(shapes)
  .then(res => console.log(res))
  .catch(err=>console.log(err))
// after 2s, the result would be [16,6,3.14]
*/


function calculateAreas(shapesObj) {
    return new Promise((resolve, reject) => {
        let result = [];

        for (let shape of shapesObj) {
            if (shape.type === "SQUARE") {
                result.push(shape.edge[0] ** 2);
            } else if (shape.type === "TRIANGLE") {
                result.push(shape.edge[0] * shape.edge[1] / 2);
            } else if (shape.type === "CIRCLE") {
                result.push(shape.edge[0] * 3.14);
            } else {
                reject((new Error(`Invalid shape type: ${shape.type}`)));
                return;
            }
        }

        setTimeout(() => {
            resolve(result);
        }, 2_000);
    });
}

calculateAreas([
    { type: "SQUARE", edge: [4] },
    { type: "TRIANGLE", edge: [3, 4] },
    { type: "CIRCLE", edge: [1] },
    // { type: "HEXAGON", "edge": [2] },
])
    .then((res) => console.log(res)) // [16, 6, 3.14]
    .catch((err) => console.log(err));

File: objects/calculate-areas.py
"""
Input:
const shapes = [
  {type: 'SQUARE', edge:[4]},
  {type: 'TRIANGLE', edge:[3,4]},
  {type: 'CIRCLE', edge: [1] },
];

Output:
calculateArea(shapes)
  .then(res => console.log(res))
  .catch(err=>console.log(err))
// after 2s, the result would be [16,6,3.14]
"""

import asyncio


async def calculate_areas(shapes_obj):
    result = []

    for shape in shapes_obj:
        if shape["shape_type"] == "SQUARE":
            result.append(shape["edge"][0] ** 2)
        elif shape["shape_type"] == "TRIANGLE":
            result.append(shape["edge"][0] * shape["edge"][1] / 2)
        elif shape["shape_type"] == "CIRCLE":
            result.append(shape["edge"][0] * 3.14)
        else:
            raise ValueError(f"Invalid shape type: {shape['shape_type']}")

    # simulate a 2-second delay (similar to setTimeout)
    await asyncio.sleep(2)
    return result


async def main():
    shapes = [
        {"shape_type": "SQUARE", "edge": [4]},
        {"shape_type": "TRIANGLE", "edge": [3, 4]},
        {"shape_type": "CIRCLE", "edge": [1]},
        # {type: "HEXAGON", "edge": [2]},
    ]

    try:
        result = await calculate_areas(shapes)
        print(result)  # output: [16, 6.0, 3.14]
    except Exception as e:
        print(f"Error: {e}")


asyncio.run(main())
